### PR TITLE
Add new `Performance/IoReadlines` cop

### PR DESCRIPTION
**Pull request by fatkodima**
`IO` (and `File` as a subclass)`readlines` class (and instance) methods read full content into memory, which can be inefficient and not necessary. This cop helps to identify such places. Specifically, when `readlines` method is followed by one of `Enumerable` methods, we can safely replace it with `each_line`, which reads one line at a time thus saving memory.

We can autocorrect usages of instance level `#readlines` method, but cannot for class level `.readlines` method due to its signature. We can only show an offense and user should manually correct it. 

```ruby
# bad
File.readlines('testfile').each { |l| puts l }
IO.readlines('testfile', chomp: true).each { |l| puts l }

conn.readlines(10).map { |l| l.size }
file.readlines.find { |l| l.start_with?('#') }
file.readlines.each { |l| puts l }

# good
File.open('testfile', 'r').each_line { |l| puts l }
IO.open('testfile').each_line(chomp: true) { |l| puts l }

conn.each_line(10).map { |l| l.size }
file.each_line.find { |l| l.start_with?('#') }
file.each_line { |l| puts l }
```

### CODE REVIEW

File: lib/rubocop/cop/performance/readlines.rb [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+      # can be replaced by `each_line` to avoid fully loading file content into memory.
+      #
+      # @example
+      #
+      #   # bad
+      #   File.readlines('testfile').each { |l| puts l }
+      #   IO.readlines('testfile', chomp: true).each { |l| puts l }
+      #
+      #   conn.readlines(10).map { |l| l.size }
+      #   file.readlines.find { |l| l.start_with?('#') }
```
Comment (koic) [inline lib/rubocop/cop/performance/readlines.rb:16]: Looks the same as the good case.

File: lib/rubocop/cop/performance/readlines.rb [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+      #
+      #   # good
+      #   File.open('testfile', 'r').each_line { |l| puts l }
+      #   IO.open('testfile').each_line(chomp: true) { |l| puts l }
+      #
+      #   conn.each_line(10).map { |l| l.size }
+      #   file.readlines.find { |l| l.start_with?('#') }
+      #   file.each_line { |l| puts l }
+      #
+      class Readlines < Cop
```
Comment (koic) [inline lib/rubocop/cop/performance/readlines.rb:27]: It's probably better to have the class specified.

```suggestion
      class IoReadlines < Cop
```

**Comment by fatkodima** [comment]
Updated.

File: config/default.yml [hunk 3]
```diff
@@ -142,6 +142,11 @@ Performance/InefficientHashSearch:
   VersionAdded: '0.56'
   Safe: false
 
+Performance/IoReadlines:
+  Description: 'Use `IO.each_line` (`IO#each_line`) instead of `IO.readlines` (`IO#readlines`).'
+  Enabled: true
```
Comment (koic) [inline config/default.yml:147]: I'm concerned about false positives when `readlines` is intentionally used as a useful shorthand. By default disabled (`Enabled: false`) seems better.

**Comment by bbatsov** [comment]
Btw, are there any references we can link to cops like this one? E.g. from Fast Ruby or something like that? I think that'd be useful to the end users who are going over the docs.

**Comment by fatkodima** [comment]
Updated: 
1. made cop disabled by default 
2. added reference (there are no anything related in fast-ruby repo, but found a good section in gitlab docs).

**Comment by koic** [comment]
Thanks!

**Pull request merged**